src/chain/chain-generic-numerator.h

```diff
@@ -98,71 +96,79 @@ class GenericNumeratorComputation {
 
-  // the alpha computation for some 0 < t <= num_time_steps_.
-  void AlphaGeneralFrame(int32 t);
+  // sets up the alpha for frame t = 0.
+  void AlphaFirstFrame(int seq);
 
-  BaseFloat ComputeTotLogLike();
+  // the alpha computation for 0 < t <= supervision_.frames_per_sequence
+  // for some 0 <= seq < supervision_.num_sequences.
+  BaseFloat AlphaGeneralFrame(int32 seq);
 
-  // sets up the beta for frame t = num_time_steps_.
-  void BetaLastFrame();
+  // the beta computation for 0 <= t < supervision_.frames_per_sequence
+  // for some 0 <= seq < supervision_.num_sequences.
+  void BetaLastFrame(int seq);
 
-  // the beta computation for 0 <= beta < num_time_steps_.
-  void BetaGeneralFrame(int32 t);
+  // the beta computation for t = supervision_.frames_per_sequence
+  void BetaGeneralFrame(int32 seq);
 
   // some checking that we can do if debug mode is activated, or on frame zero.
   // Sets ok_ to false if a bad problem is detected.
-  void BetaGeneralFrameDebug(int32 t);
+  void BetaGeneralFrameDebug(int32 seq);
 
 
   const Supervision &supervision_;
 
-  // the transposed neural net output.
-  Matrix<BaseFloat> exp_nnet_output_transposed_;
+  // a reference to the nnet output.
+  const CuMatrixBase<BaseFloat> &nnet_output_x_;
 
   // in_transitions_ lists all the incoming transitions for
   // each state of each numerator graph
   // out_transitions_ does the same but for the outgoing transitions
-  std::vector<std::vector<std::vector<DenominatorGraphTransition> > >
-  in_transitions_, out_transitions_;
+  typedef std::vector<std::vector<DenominatorGraphTransition> > TransitionMap;
+  std::vector<TransitionMap> in_transitions_, out_transitions_;
+  std::vector<std::vector<MatrixIndexT>> index_to_pdf_;
 
   // final probs for each state of each numerator graph
-  Matrix<double> final_probs_; // indexed by seq, state
+  Matrix<BaseFloat> final_probs_;  // indexed by seq, state
 
   // an offset subtracted from the logprobs of transitions out of the first
   // state of each graph to help reduce numerical problems. Note the
   // generic forward-backward computations cannot be done in log-space.
   Vector<BaseFloat> offsets_;
 
-  // maximum number of states among all the numerator graphs
-  // (it is used as a stride in alpha_ and beta_)
-  int32 max_num_hmm_states_;
-
   // the derivs w.r.t. the nnet outputs (transposed)
   // (the dimensions and functionality is the same as in
   // DenominatorComputation)
-  Matrix<BaseFloat> nnet_output_deriv_transposed_;
+  std::vector<Matrix<BaseFloat>> nnet_output_deriv_;
 
   // forward and backward probs matrices. These have the
   // same dimension and functionality as alpha_ and beta_
   // in DenominatorComputation except here we don't use beta
   // sums (becasue we don't use leakyHMM). However, we use
   // alpha sums to help avoid numerical issues.
-  Matrix<double> alpha_;
-  Matrix<double> beta_;
+  std::vector<Matrix<BaseFloat>> alpha_;
```

Comment:
Is it necessary to have  these things as class members instead of as local variables inside functions (possibly passed between functions as necessary)?  It seems wasteful in terms of memory, although I know in the big picture it is trivial.

Reply:
This is the only part that can take more time -- for now, I didn't change the way this class is used (in chain/chain-training.cc), i.e. the interface is the same. 

I can change it so that we can have only one Matrix as a member variables (instead of vector<Matrix>) or even make a new member function (say with signature ForwardBackward(int sequence_id, CuMatrix *out)) which will have only local variables alpha, beta and probs (they will need to be allocated for each sequence again).